File: _book/source/_posts/Algorithm/1.leetcode/tmp/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#
from collections import defaultdict


# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        ans = 0
        counter = defaultdict(lambda: 0)
        for r in range(len(s)):
            # 获取当前元素对应的位置
            while counter.get(s[r], 0) != 0:
                # 如果当前遍历字符出现过，
                # 缩小窗口，获取上一次出现的索引位置，
                counter[s[l]] = counter.get(s[l], 0) - 1
                l += 1
            counter[s[r]] += 1
            ans = max(ans, r - l + 1)
        return ans


test = Solution()
test.lengthOfLongestSubstring("abccbcbb")

# @lc code=end

File: source/_posts/Algorithm/算法专题/滑动窗口/3.无重复字符的最长子串.java

import java.util.HashSet;

/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 * 一旦发现set存在，从左侧i开始去除元素，直到不再重复
 */

// @lc code=start
class Solution {

    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> hashSet = new HashSet<>();
        int j = 0;
        int i = 0;
        int res = 0;
        while (j < s.length()) {
            if (!hashSet.contains(s.charAt(j))) {
                hashSet.add(s.charAt(j++));
                res = Math.max(hashSet.size(), res);
            } else {
                hashSet.remove(s.charAt(i++));
            }
        }
        return res;

    }

    // public static void main(String[] args) {
    // Solution3 solution = new Solution3();
    // solution.lengthOfLongestSubstring(" ");
    // }

}
// @lc code=end
